src/main/java/pingis/controllers/LiveChallengeController.java

```diff
@@ -261,16 +291,75 @@ public RedirectView playArcade(RedirectAttributes redirectAttributes,
       redirectAttributes.addFlashAttribute("minLength", Integer.MAX_VALUE);
       logger.info("User has test turn. Redirecting to /newtaskpair.");
       return new RedirectView("/newtaskpair");
+
     } else {
       Task implTask = gameplayService.getRandomImplementationTask(challenge);
       Task testTask = taskService.getCorrespondingTask(implTask);
-      redirectAttributes.addAttribute("taskId", implTask.getId());
-      redirectAttributes.addAttribute("testTaskInstanceId",
-          taskInstanceService.getByTaskAndUser(testTask, testTask.getAuthor()).getId());
-      logger.info("User has implementation turn. Redirecting to /newTaskInstance");
-      return new RedirectView("/newTaskInstance");
+      TaskInstance testTaskInstance =
+          taskInstanceService.getByTaskAndUser(testTask, testTask.getAuthor());
+      logger.info("User has implementation turn.");
+      return newTaskInstance(implTask, testTaskInstance, redirectAttributes);
+    }
+  }
+
+  private RedirectView playPractice(RedirectAttributes redirectAttributes, Challenge challenge) {
+    // check for unfinished instances
+    User player = userService.getCurrentUser();
+    logger.info("User: " + player.getName());
+
+    if (challenge.getAuthor().equals(player)
+        || (challenge.getSecondPlayer() != null
+            && challenge.getSecondPlayer().equals(player))) {
```

Comment:
It might be cleaner to move some of these multi-part booleans a separate service method. There could even be a separate "turn service" for figuring things like this out. Just an idea though, it could be unnecessary or more confusing.